File: 2021/09/02/IslandPerimeter.java
/**
 * https://leetcode.com/problems/island-perimeter/
 * @author Andrew Jarombek
 * @since 9/2/2021
 */

public class IslandPerimeter {
    public static int islandPerimeter(int[][] grid) {
        int perimeter = 0;

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                if (grid[i][j] == 1) {
                    if (i-1 < 0 || grid[i-1][j] == 0) {
                        perimeter++;
                    }

                    if (j-1 < 0 || grid[i][j-1] == 0) {
                        perimeter++;
                    }

                    if (i+1 == grid.length || grid[i+1][j] == 0) {
                        perimeter++;
                    }

                    if (j+1 == grid[i].length || grid[i][j+1] == 0) {
                        perimeter++;
                    }
                }
            }
        }

        return perimeter;
    }

    public static void main(String[] args) {
        assert islandPerimeter(new int[][]{new int[] {0, 1}}) == 4;
    }
}

File: 2022/03/21/islandPerimeter.ts
/**
 * https://leetcode.com/problems/island-perimeter/
 * @author Andrew Jarombek
 * @since 3/21/2022
 */

function islandPerimeter(grid: number[][]): number {
    let i = 0;
    let j = 0;
    let found = false;

    while (!found && i < grid.length) {
        while (!found && j < grid[0].length) {
            if (grid[i][j] == 1) {
                found = true;
                return searchIsland(grid, i, j);
            }

            j++;
        }

        i++;
        j = 0;
    }

    return 0;
}

function searchIsland(grid: number[][], i: number, j: number): number {
    let value = 0;
    grid[i][j] = 2;

    if (i != 0 && grid[i-1][j] == 1) {
        value += searchIsland(grid, i-1, j);
    } else if (i == 0 || grid[i-1][j] == 0) {
        value++;
    }

    if (i != grid.length - 1 && grid[i+1][j] == 1) {
        value += searchIsland(grid, i+1, j);
    } else if (i == grid.length - 1 || grid[i+1][j] == 0) {
        value++;
    }

    if (j != 0 && grid[i][j-1] == 1) {
        value += searchIsland(grid, i, j-1);
    } else if (j == 0 || grid[i][j-1] == 0) {
        value++;
    }

    if (j != grid[0].length - 1 && grid[i][j+1] == 1) {
        value += searchIsland(grid, i, j+1);
    } else if (j == grid[0].length - 1 || grid[i][j+1] == 0) {
        value++;
    }

    return value;
}